object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/CalledMethodsPredicateEvaluator.java

```diff
@@ -1,23 +1,181 @@
+  private static SolverContext setupSolver() {
+    Configuration config = Configuration.defaultConfiguration();
+    LogManager log = LogManager.createNullLogManager();
+    ShutdownNotifier notifier = ShutdownManager.create().getNotifier();
+    try {
+      return SolverContextFactory.createSolverContext(
+          config, log, notifier, SolverContextFactory.Solvers.SMTINTERPOL);
+    } catch (InvalidConfigurationException e) {
+      return null;
+    }
+  }
+
+  /**
+   * Return true iff the boolean formula "lhs &rarr; rhs" is valid, treating all variables as
+   * uninterpreted.
+   *
+   * @param lhs a formula in Java format as a String containing only {@literal &}{@literal &}, ||,
+   *     (), !, and literals
+   * @param rhs a formula in Java format as a String containing only {@literal &}{@literal &}, ||,
+   *     (), !, and literals
+   * @return true iff lhs implies rhs
+   * @throws UnsupportedOperationException if the input contains an unexpected kind of expression
+   */
+  public static boolean implies(final String lhs, final String rhs) {
+
+    /**
+     * General approach: parse each formula into an AST, then convert each AST into a boolean
+     * formula that can be passed to a solver, and then construct a query to check if "not (lhs ->
+     * rhs)" is unsatisfiable. The result is the result of that query.
+     */
+
+    // setup solver
+    SolverContext context = setupSolver();
+    BooleanFormulaManager booleanFormulaManager =
+        context.getFormulaManager().getBooleanFormulaManager();
+
+    // parse the formulas into the solver's format
+    BooleanFormula lhsBool = formulaStringToBooleanFormula(lhs, booleanFormulaManager);
+    BooleanFormula rhsBool = formulaStringToBooleanFormula(rhs, booleanFormulaManager);
+
+    BooleanFormula satQuery =
+        booleanFormulaManager.not(booleanFormulaManager.implication(lhsBool, rhsBool));
+    ProverEnvironment prover =
+        context.newProverEnvironment(SolverContext.ProverOptions.GENERATE_MODELS);
+
+    try {
+      prover.addConstraint(satQuery);
+    } catch (InterruptedException e) {
+      return false;
+    }
+
+    try {
+      return prover.isUnsat();
+    } catch (Exception e) {
+      return false;
+    }
+  }
+
+  /**
+   * Converts a String representing a boolean expression into a boolean formula in the SMT solver's
+   * format.
+   */
+  private static BooleanFormula formulaStringToBooleanFormula(
+      String formula, BooleanFormulaManager booleanFormulaManager) {
+
+    String classWithFormula = "class DUMMY { boolean DUMMY() { return " + formula + "; } }";
+
+    CompilationUnit ast = parser.parse(classWithFormula).getResult().orElse(null);
+
+    if (ast == null) {
+      throw new BugInCF(
+          "Encountered an unparseable formula while parsing an @CalledMethodsPredicate "
+              + "annotation, but ObjectConstructionVisitor failed to stop compilation. Unparseable formula: "
```

Comment:
The fact about ObjectConstructionVisitor is an implementation detail.  You could write a code comment here, but I would keep the error message short (just the last, brief sentence).